sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,681 @@
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+
+    Examples
+    --------
+    Adding a Pin Joint which connects center of mass of parent to a point
+    pointed by child.frame.x + child.frame.y w.r.t. child's center of mass.
+    Gravity is along y axis of parent.
+
+    >>> from sympy import Symbol
+    >>> from sympy.physics.mechanics import Body, PinJoint
+    >>> parent = Body('parent')
+    >>> child = Body('child')
+    >>> gravity = Symbol('gravity')
+    >>> l = Symbol('l')
+    >>> child.apply_force(child.mass * gravity * parent.frame.y, child.masscenter)
+    >>> pin_joint = PinJoint('pinjoint', parent, \
+                             child, child_point_pos=(l, l, 0))
+    >>> pin_joint.coordinates
+    [pinjoint_theta(t)]
+    >>> pin_joint.speeds
+    [pinjoint_omega(t)]
+
+    """
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None, parent_axis=None, child_axis=None):
+
+        parent_axes_str = {'X': parent.frame.x, 'Y': parent.frame.y,
+                           'Z': parent.frame.z}
+        child_axes_str = {'X': child.frame.x, 'Y': child.frame.y,
+                          'Z': child.frame.z}
+
+        if parent_axis is None:
+            self.parent_axis = parent.frame.x
+        elif isinstance(parent_axis, tuple):
+            self.parent_axis = convert_tuple_to_vector(self.parent.frame, parent_axis)
+        elif isinstance(parent_axis, str) and parent_axis.upper() \
+                in parent_axes_str.keys():
+            self.parent_axis = parent_axes_str[parent_axis.upper()]
+        else:
+            raise TypeError("Parent Axis must either be one of 'X', 'Y', 'Z'" +
+                            " or a 3-Tuple.")
+
+        if child_axis is None:
+            self.child_axis = child.frame.x
+        elif isinstance(child_axis, tuple):
+            self.child_axis = convert_tuple_to_vector(self.child.frame, child_axis)
+        elif isinstance(child_axis, str) and child_axis.upper() \
+                in child_axes_str.keys():
+            self.child_axis = child_axes_str[child_axis.upper()]
+        else:
+            raise TypeError("Child Axis must either be one of 'X', 'Y', 'Z'" +
+                            " or a 3 Tuple")
+
+        super(PinJoint, self).__init__(name, parent, child, parent_point_pos,
+                                       child_point_pos)
+
+    def apply_joint(self):
+        theta = dynamicsymbols(self.name + '_theta')
+        thetad = dynamicsymbols(self.name + '_theta', 1)
+        omega = dynamicsymbols(self.name + '_omega')
```

Comment:
@moorepants Shouldn't we allow users to explicitly pass symbols for theta and omega to __init__()?

Reply:
Yes, making these variables configurable and accessible is important.